examples/rgcn_link_prediction.py

```diff
@@ -0,0 +1,232 @@
+from tqdm import tqdm
+import numpy as np
+import torch
+import torch.nn.functional as F
+import torch.nn as nn
+from torch.utils.tensorboard import SummaryWriter
+from torch_geometric.datasets.link_prediction import LinkPrediction
+from torch_geometric.nn import GAE, RGCNConv
+from torch_geometric.utils.evaluate_relational_link_prediction import evaluate_relational_link_prediction
+
+""""
+Implements link prediction on the FB15k237 datasets according to
+the `"Modeling Relational Data with Graph Convolutional Networks"
+<https://arxiv.org/abs/1703.06103>`_ paper.
+Caution: very high memory consumption. You should have at least 64GB RAM in your machine. 
+"""
+
+
+class RGCNEncoder(torch.nn.Module):
+    def __init__(self, in_channels, out_channels, num_relations):
+        super(RGCNEncoder, self).__init__()
+        self.linear_features = nn.Linear(in_channels, 500)
+        self.conv1 = RGCNConv(500, out_channels, num_relations=num_relations, num_blocks=5)
+        self.conv2 = RGCNConv(out_channels, out_channels, num_relations=num_relations, num_blocks=5)
+
+    def forward(self, x, edge_index, edge_type):
+        x = self.linear_features(x)
+        x = F.relu(self.conv1(x, edge_index, edge_type))
+        x = F.dropout(x, p=0.2, training=self.training)
+        x = self.conv2(x, edge_index, edge_type)
+        return x
+
+
+class DistMultDecoder(torch.nn.Module):
+    def __init__(self, num_relations, relation_embedding_dim):
+        super(DistMultDecoder, self).__init__()
+        self.relation_embedding = nn.Parameter(torch.Tensor(num_relations, relation_embedding_dim))
+        nn.init.xavier_uniform_(self.relation_embedding, gain=nn.init.calculate_gain('relu'))
+
+    def forward(self, z, edge_index, edge_type):
+        s = z[edge_index[0, :]]
+        r = self.relation_embedding[edge_type]
+        o = z[edge_index[1, :]]
+        score = torch.sum(s * r * o, dim=1)
+        return score
+
+
+def negative_sampling(edge_index, edge_type, num_nodes, num_edges, omega=1.0, corrupt_subject=True,
+                      corrupt_predicate=False, corrupt_object=True):
+    negative_samples_count = int(omega * edge_type.size(0))
+    permutation = np.random.choice(np.arange(edge_type.size(0)), negative_samples_count)
+    s = edge_index[0][permutation]
+    r = edge_type[permutation]
+    o = edge_index[0][permutation]
```

Comment:
Hope it's ok if I comment. Should it be `o = edge_index[1][permutation]`?

Reply:
Oh yeah, you're right. Thanks!